File: src/sudoku.rs
#[cfg(feature = "std")]
use std::{
    fmt,
    num::NonZeroU8,
    io::{self, Read},
};
#[cfg(not(feature = "std"))]
use core::{
    fmt,
    num::NonZeroU8,
};

// The width and height of a single "box" on the board
const BOX_SIZE: usize = 3; // 3 cells per box
// The width and height of the entire board
const BOARD_SIZE: usize = 3 * BOX_SIZE; // 3 boxes

#[derive(Debug, Clone)]
pub enum SolverError {
    NoSolution,
}

#[cfg(feature = "std")]
#[derive(Debug)]
pub enum ReadError {
    IOError(io::Error),
    ExpectedNewline,
    InvalidDigit(char),
}

#[derive(Debug, Clone, Copy, Eq)]
pub struct Tile {
    /// The numeric value of the tile - value between 1 and BOARD_SIZE (inclusive)
    /// None - tile is empty
    value: Option<NonZeroU8>,

    /// Represents every possible value that could be placed on this tile
    ///
    /// Each position is associated with a value in the range 1 to BOARD_SIZE+1
    /// true if a value is available, false if a value cannot be used on this tile
    possible_values: [bool; BOARD_SIZE],

    /// A cache of the number of true values in possible_values
    /// Not used if tile is not empty
    possible_count: usize,
}

impl Default for Tile {
    fn default() -> Self {
        // While we can't guarantee that an empty tile can have anything placed on it in every
        // situation, implementing it this way saves us from having to have a more complicated
        // implementation of Default for Sudoku
        Tile {
            value: None,
            possible_values: [true; BOARD_SIZE],
            possible_count: BOARD_SIZE,
        }
    }
}

impl fmt::Display for Tile {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        match self.value {
            Some(value) => write!(f, "{}", value),
            None => write!(f, "0"),
        }
    }
}

impl PartialEq for Tile {
    fn eq(&self, other: &Self) -> bool {
        self.value == other.value
    }
}

impl Tile {
    pub fn is_empty(&self) -> bool {
        self.value.is_none()
    }

    pub fn set(&mut self, value: NonZeroU8) {
        self.value = Some(value);
    }

    /// Marks the given value as no longer available for this tile
    pub fn mark_unavailable(&mut self, value: NonZeroU8) {
        // The index at which this value will be found in all possible_values array
        // This subtraction can never panic because NonZeroU8 >= 1
        let value_index = value.get() as usize - 1;
        if self.possible_values[value_index] {
            // Tile was available, but now it is not
            self.possible_values[value_index] = false;
            self.possible_count -= 1;
        }
    }
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Sudoku {
    /// Tiles are stored row-wise
    /// So index with tiles[row][col]
    tiles: [[Tile; BOARD_SIZE]; BOARD_SIZE],
    /// The number of empty tiles remaining
    empty_tiles: usize,
}

impl Default for Sudoku {
    fn default() -> Self {
        Sudoku {
            tiles: Default::default(),
            empty_tiles: BOARD_SIZE * BOARD_SIZE,
        }
    }
}

impl fmt::Display for Sudoku {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        for row in &self.tiles {
            for tile in row {
                write!(f, "{}", tile)?;
            }
            writeln!(f)?;
        }

        Ok(())
    }
}

impl Sudoku {
    /// Creates a new sudoku board by reading 9 rows of 9 digits representing the values currently
    /// filled and unfilled in the grid. 0 represents an empty tile.
    #[cfg(feature = "std")]
    pub fn read_board<R: Read>(mut reader: R) -> Result<Self, ReadError> {
        debug_assert_eq!(BOARD_SIZE, 9, "This function assumes that BOARD_SIZE is 9");

        let mut board = Self::default();

        // Read each value + newline
        let mut buffer = [0; BOARD_SIZE + 1];
        for row_i in 0..BOARD_SIZE {
            reader.read_exact(&mut buffer).map_err(ReadError::IOError)?;
            if buffer[BOARD_SIZE] != b'\n' {
                return Err(ReadError::ExpectedNewline);
            }

            for col_i in 0..BOARD_SIZE {
                let value = buffer[col_i];
                if value.is_ascii_digit() {
                    // Only place the tile if the value is not zero
                    NonZeroU8::new(value - b'0').map(|value| {
                        board.place((row_i, col_i), value);
                    });
                }
                else {
                    return Err(ReadError::InvalidDigit(value as char));
                }
            }
        }

        Ok(board)
    }

    /// Returns true if the board is a valid sudoku board
    /// A board is invalid if there are duplicate numbers in any row, column, or box
    /// or if any of its values are out of the range 0 to BOARD_SIZE inclusive.
    pub fn is_valid(&self) -> bool {
        true // TODO
    }

    /// Returns true if the board is completely filled
    pub fn is_complete_board(&self) -> bool {
        self.empty_tiles == 0
    }

    /// Provides a rating of the difficulty of a puzzle based on how many empty tiles there are
    pub fn difficulty(&self) -> f64 {
        1.0 - self.empty_tiles as f64 / (BOARD_SIZE*BOARD_SIZE) as f64
    }

    /// Places a value on the sudoku board. Updates all related possible value
    /// caches and their counts.
    pub fn place(&mut self, (row, col): (usize, usize), value: NonZeroU8) {
        debug_assert!(value.get() <= BOARD_SIZE as u8, "value `{}` is invalid", value);
        // Only decrement the empty tile count if the tile was previously empty
        if self.tiles[row][col].is_empty() {
            self.empty_tiles -= 1;
        }
        self.tiles[row][col].set(value);

        // The start index of this tile's box
        let box_row_start = (row / BOX_SIZE) * BOX_SIZE;
        let box_col_start = (col / BOX_SIZE) * BOX_SIZE;

        for i in 0..BOARD_SIZE {
            // Update items in the same row as the placed tile
            self.tiles[row][i].mark_unavailable(value);

            // Update items in the same column as the placed tile
            self.tiles[i][col].mark_unavailable(value);

            // Update items in the same box as the placed tile
            let box_row = box_row_start + i / BOX_SIZE;
            let box_col = box_col_start + i % BOX_SIZE;
            self.tiles[box_row][box_col].mark_unavailable(value);
        }
    }

    /// Attempts to solve the sudoku board
    ///
    /// Only fails if all possible solutions are exhausted
    pub fn solve(&mut self) -> Result<(), SolverError> {
        if self.is_complete_board() {
            Ok(())
        }
        else {
            self.guess_solve()
        }
    }

    /// Second tier of solving (intelligent brute force)
    /// Makes an intelligent guess about what position to guess on
    /// Guesses an appropriate value and attempts to solve from the first tier
    /// using the modified board
    /// If no solution is found, the initial guess is undone and the search
    /// continues
    /// Only fails if all possible solutions are exhausted.
    fn guess_solve(&mut self) -> Result<(), SolverError> {
        // Get the tile with the minimum number of possibilities
        // This is the most efficient place to start guessing because
        // if we guess wrong we will have the fewest number of alternatives
        // to search
        // On the otherhand, if we chose the tile with the maximum number
        // of possible values, we elimate more possible routes by guessing
        // right. I'm just betting that we'll guess wrong more often then
        // we guess right since there are more wrong numbers than right ones.

        let (min_row, min_col) = self.min_possible_empty_tile()?;
        let possible_values = self.tiles[min_row][min_col].possible_values;

        for i in 0..BOARD_SIZE {
            if !possible_values[i] {
                // Skip values that cannot possibly be solutions
                continue;
            }

            // The value at this index is to be used as the guess
            // Each index in possible_values represents a value from 1-9
            // Adding 1 to the index produces the value
            // This unwrap is safe because a u8 is >= 0 and adding 1 means the value is >= 1
            let guess = NonZeroU8::new(i as u8 + 1).unwrap();

            // Make a guess
            let mut copy = self.clone();
            copy.place((min_row, min_col), guess);

            // Try to solve the board with this guess
            if copy.solve().is_ok() {
                *self = copy;
                return Ok(())
            }
        }

        Err(SolverError::NoSolution)
    }

    // Returns the tile with the minimum number of possibilities
    fn min_possible_empty_tile(&self) -> Result<(usize, usize), SolverError> {
        let mut min = Err(SolverError::NoSolution);

        for (row_i, row) in self.tiles.iter().enumerate() {
            for (col_i, tile) in row.iter().enumerate() {
                // If the tile is already filled, move on
                if !tile.is_empty() {
                    continue;
                }

                min = match min {
                    Err(_) => Ok((row_i, col_i, tile.possible_count)),
                    Ok((_, _, possible_count)) if tile.possible_count < possible_count => {
                        Ok((row_i, col_i, tile.possible_count))
                    },
                    _ => min,
                }
            }
        }

        min.map(|(row, col, _)| (row, col))
    }
}

File: c-impl/sudoku.c
/**
 * Sudoku Board Primitives
 *
 * Author: Sunjay Varma (www.sunjay.ca)
 */
#include <stdbool.h>

#include "sudoku.h"

/**
 * Converts tile column/row indexes to the appropriate index in the
 * SudokuBoard tiles property
 */
static int coordinatesToTileIndex(int row_i, int col_i) {
    return row_i * BOARD_SIZE + col_i;
}

/**
 * Initializes a sudoku board to be a completely empty (all zeros) board
 * Does not allocate any memory
 */
void emptySudokuBoard(SudokuBoard* board) {
    int index;

    for (int i = 0; i < BOARD_SIZE; i++) {
        for (int j = 0; j < BOARD_SIZE; j++) {
            index = coordinatesToTileIndex(i, j);
            // There are BOARD_SIZE possible values for every tile on
            // an empty board
            for (int p = 0; p < BOARD_SIZE; p++) {
                board->tiles[index].possibleValues[p] = true;
            }
            board->tiles[index].possibleCount = BOARD_SIZE;
        }
    }
}

/**
 * Copies a board and all of its values to the provided copy
 * Does not allocate any memory
 */
void copySudokuBoard(SudokuBoard* board, SudokuBoard* copy) {
    // Let the compiler generate the copy code
    *copy = *board;
}

/**
 * Places a value on the sudoku board. Updates all related possible value
 * caches and their counts.
 */
void placeSudokuValue(SudokuBoard* board, int row_i, int col_i, short value) {
    // Place the value on its tile
    int index = coordinatesToTileIndex(row_i, col_i);
    board->tiles[index].value = value;

    if (value == 0) {
        return;
    }

    // The start index of this tile's box
    int boxRowStart = (row_i / BOX_SIZE) * BOX_SIZE;
    int boxColStart = (col_i / BOX_SIZE) * BOX_SIZE;

    // The index at which this value will be found in all possibleValue arrays
    int valueIndex = value - 1;

    for (int i = 0; i < BOARD_SIZE; i++) {
        // Update items in the same row
        index = coordinatesToTileIndex(row_i, i);
        Tile* rowItem = &(board->tiles[index]);
        if (rowItem->possibleValues[valueIndex]) {
            // Tile was available, now it is not
            rowItem->possibleValues[valueIndex] = false;
            rowItem->possibleCount--;
        }

        // Update items in the same column
        index = coordinatesToTileIndex(i, col_i);
        Tile* colItem = &(board->tiles[index]);
        if (colItem->possibleValues[valueIndex]) {
            // Tile was available, now it is not
            colItem->possibleValues[valueIndex] = false;
            colItem->possibleCount--;
        }

        // Update items in the same box
        index = coordinatesToTileIndex(boxRowStart + i / BOX_SIZE,
                    boxColStart + i % BOX_SIZE);
        Tile* boxItem = &(board->tiles[index]);
        if (boxItem->possibleValues[valueIndex]) {
            // Tile was available, now it is not
            boxItem->possibleValues[valueIndex] = false;
            boxItem->possibleCount--;
        }
    }
}

/**
 * Sets all items in a single board row to items
 */
void setBoardRow(SudokuBoard* board, int row_i, Tile items[BOARD_SIZE]) {
    int index;

    for (int i = 0; i < BOARD_SIZE; i++) {
        index = coordinatesToTileIndex(row_i, i);
        board->tiles[index] = items[i];
    }
}

/**
 * Sets all values of all items in a single board row to items
 */
void setBoardRowValues(SudokuBoard* board, int row_i, short items[BOARD_SIZE]) {
    for (int i = 0; i < BOARD_SIZE; i++) {
        placeSudokuValue(board, row_i, i, items[i]);
    }
}

/**
 * Gets a single Tile at the given row_i and col_i position
 * row_i and col_i are indexes where 0 <= i < BOARD_SIZE
 * Tile is a pointer to a single Tile
 */
void getBoardTile(SudokuBoard* board, int row_i, int col_i, Tile* tile) {
    int index = coordinatesToTileIndex(row_i, col_i);
    // The compiler will generate appropriate copy code
    *tile = board->tiles[index];
}

/**
 * Copies the values from the row at the given index into the given row
 * Performs no allocation. row must be large enough for BOARD_SIZE Tiles
 */
void getBoardRow(SudokuBoard* board, int row_i, Tile row[BOARD_SIZE]) {
    for (int i = 0; i < BOARD_SIZE; i++) {
        getBoardTile(board, row_i, i, &row[i]);
    }
}

/**
 * Copies the values from the column at the given index into the given col
 * Performs no allocation. col must be large enough for BOARD_SIZE Tiles
 */
void getBoardColumn(SudokuBoard* board, int col_i, Tile col[BOARD_SIZE]) {
    for (int i = 0; i < BOARD_SIZE; i++) {
        getBoardTile(board, i, col_i, &col[i]);
    }
}

/**
 * Copies the values of the given "box index" into the provided box
 * box must be an array of array pointers that represents the box
 * Does not allocate any memory
 *
 * Box indexes work like this for BOARD_SIZE = 9:
 * 0 1 2
 * 3 4 5
 * 6 7 8
 */
void getBoardBox(SudokuBoard* board, int box_i, Tile box[BOX_SIZE][BOX_SIZE]) {
    int y_offset = (box_i / ((int)BOX_SIZE)) * BOX_SIZE;
    int x_offset = (box_i % ((int)BOX_SIZE)) * BOX_SIZE;
    for (int i = 0; i < BOX_SIZE; i++) {
        for (int j = 0; j < BOX_SIZE; j++) {
            getBoardTile(board, y_offset + i, x_offset + j, &box[i][j]);
        }
    }
}

/**
 * Checks for duplicates
 */
static bool isValid(Tile* row) {
    bool found[BOARD_SIZE] = {false, false, false, false, false, false, false, false, false};
    for (int i = 0; i < BOARD_SIZE; i++) {
        short value = row[i].value;
        if (value == 0) {
            continue;
        }
        if (found[value-1]) {
            return false;
        }
        found[value-1] = true;
    }
    return true;
}

/**
 * Validates whether the board is correct
 *
 * A board is incorrect if it has any repetitions
 */
bool isValidBoard(SudokuBoard* board) {
    Tile row[BOARD_SIZE];
    Tile col[BOARD_SIZE];
    Tile box[BOX_SIZE][BOX_SIZE];

    for (int i = 0; i < BOARD_SIZE; i++) {
        // validate box
        getBoardBox(board, i, box);

        Tile box_values[BOARD_SIZE];
        for (int i = 0; i < BOX_SIZE; i++) {
            for (int j = 0; j < BOX_SIZE; j++) {
                box_values[i*BOX_SIZE + j] = box[i][j];
            }
        }

        if (!isValid(box_values)) {
            return false;
        }

        // validate column
        getBoardColumn(board, i, col);
        bool colValid = isValid(col);

        if (!colValid) {
            return false;
        }

        // validate row
        getBoardRow(board, i, row);
        bool rowValid = isValid(row);

        if (!rowValid) {
            return false;
        }
    }
    return true;
}

/**
 * Returns whether you have a valid, complete board
 */
bool isCompleteBoard(SudokuBoard* board) {
    int index;

    // make sure everything is full
    for (int i = 0; i < BOARD_SIZE; i++) {
        for (int j = 0; j < BOARD_SIZE; j++) {
            index = coordinatesToTileIndex(i, j);

            if (board->tiles[index].value == 0) {
                return false;
            }
        }
    }

    return isValidBoard(board);
}

/**
 * Rates the given board on a scale from 0.0 to 1.0
 * based on how difficult it is.
 */
double getBoardDifficultyRating(SudokuBoard* board) {
    int index;

    // More empty tiles = more difficult
    double empty = 0;
    // Loop through all tiles and count how many are already filled
    for (int i = 0; i < BOARD_SIZE; i++) {
        for (int j = 0; j < BOARD_SIZE; j++) {
            index = coordinatesToTileIndex(i, j);

            if (board->tiles[index].value == 0) {
                empty += 1;
            }
        }
    }
    
    // Returns the inverse divided by the maximum board size
    return empty / (BOARD_SIZE * BOARD_SIZE);
}

